runtime/src/chpl-cache.c

```diff
@@ -3566,13 +3566,33 @@ static struct rdcache_s* cache_create(void);
 CHPL_TLS_DECL(struct rdcache_s*,cache_remote_data);
 static pthread_key_t pthread_cache_info_key; // stores struct rdcache_s*
 
+// ideally we would just make is_inited an atomic but
+// the runtime atomics support can't handle static initialization
+// so we use a pthread mutex for now
+static pthread_mutex_t is_inited_mutex = PTHREAD_MUTEX_INITIALIZER;
+static chpl_bool is_inited = 0; // protected by is_inited_mutex
+
+// returns NULL if the cache is not yet inited
 static
 struct rdcache_s* tls_cache_remote_data(void) {
   struct rdcache_s *cache = CHPL_TLS_GET(cache_remote_data);
```

Comment:
In testing https://github.com/chapel-lang/chapel/pull/18802 I discovered this is broken on platforms where we use `pthread_key_t` to implement TLS. We end up calling `pthread_getspecific` before we've called `pthread_key_create`, which isn't legal. From https://linux.die.net/man/3/pthread_getspecific

> The effect of calling pthread_getspecific() or pthread_setspecific() with a key value not obtained from pthread_key_create() or after key has been deleted with pthread_key_delete() is undefined.

See also https://pubs.opengroup.org/onlinepubs/009695399/functions/pthread_key_create.html for more context and an example of what we're doing wrong.

Fixing that with a `pthread_once` init, I'm seeing failures at shutdown during a barrier in `chpl_comm_exit()`. Not sure what that's about yet.

Reply:
So you're saying my thanks was premature?  :D

Reply:
Debugged with Michael, fix will be in https://github.com/chapel-lang/chapel/pull/18802